osu.Game/Graphics/Containers/SectionsContainer.cs

```diff
@@ -183,72 +183,82 @@ protected override void UpdateAfterChildren()
 
-                if (ExpandableHeader != null && FixedHeader != null)
-                {
-                    float offset = Math.Min(ExpandableHeader.LayoutSize.Y, currentScroll);
+                ExpandableHeader.Y = -offset;
+                FixedHeader.Y = -offset + ExpandableHeader.LayoutSize.Y;
+            }
 
-                    ExpandableHeader.Y = -offset;
-                    FixedHeader.Y = -offset + ExpandableHeader.LayoutSize.Y;
-                }
+            headerBackgroundContainer.Height = (ExpandableHeader?.LayoutSize.Y ?? 0) + (FixedHeader?.LayoutSize.Y ?? 0);
+            headerBackgroundContainer.Y = ExpandableHeader?.Y ?? 0;
 
-                headerBackgroundContainer.Height = (ExpandableHeader?.LayoutSize.Y ?? 0) + (FixedHeader?.LayoutSize.Y ?? 0);
-                headerBackgroundContainer.Y = ExpandableHeader?.Y ?? 0;
+            selectMostVisible();
+        }
 
-                float scrollOffset = FixedHeader?.LayoutSize.Y ?? 0;
+        /// <summary>
+        /// Changes <see cref="SelectedSection"/> to currently most visible section
+        /// </summary>
+        private void selectMostVisible()
+        {
+            // reset last clicked section because user started scrolling themselves
+            if (scrollContainer.UserScrolling)
+                lastClickedSection = null;
 
-                float visibleHeight(T section)
-                {
-                    float sectionPosition = scrollContainer.GetChildPosInContent(section) - scrollOffset;
-                    float top = Math.Max(sectionPosition, currentScroll);
-                    float bottom = Math.Min(sectionPosition + section.Height, currentScroll + scrollContainer.DisplayableContent);
-                    return bottom - top;
-                }
+            // we are scrolled all the way to the bottom
+            // select the section user clicked for or the very last section
+            if (Precision.AlmostBigger(scrollContainer.Current, scrollContainer.ScrollableExtent))
+            {
+                SelectedSection.Value = lastClickedSection as T ?? Children.LastOrDefault();
+                return;
+            }
 
-                var sortedByVisible = new List<T>(Children);
-                sortedByVisible.Sort((a, b) => visibleHeight(b).CompareTo(visibleHeight(a)));
-                T mostVisible = sortedByVisible.FirstOrDefault();
+            var sortedByVisibility = new List<T>(Children);
+            sortedByVisibility.Sort((a, b) => visibleHeight(b).CompareTo(visibleHeight(a)));
+            T mostVisible = sortedByVisibility.FirstOrDefault();
 
-                if (mostVisible == null)
-                    return;
+            if (mostVisible == null)
+                return;
 
-                int mostVisibleIndex = IndexOf(mostVisible);
+            int mostVisibleIndex = IndexOf(mostVisible);
+            while (isPreviousMoreVisible(mostVisibleIndex))
```

Comment:
i think this may read better if you make it into a `for` loop with a `break` condition.

Reply:
I think the best thing to do is actually rewrite the sort. Sort on this "more visible or 100% visible" notion to begin with. 

Got this idea while driving to work, I'll implement it when I get home.